File: src/main/java/com/github/wilgaboury/sigwig/Button.java
package com.github.wilgaboury.sigwig;

import com.github.wilgaboury.jsignal.Signal;
import com.github.wilgaboury.sigui.*;
import io.github.humbleui.skija.Canvas;
import io.github.humbleui.skija.Paint;
import io.github.humbleui.types.Rect;
import org.lwjgl.util.yoga.Yoga;

import java.util.function.Supplier;

import static com.github.wilgaboury.jsignal.ReactiveUtil.*;
import static com.github.wilgaboury.sigui.event.EventListener.*;
import static com.github.wilgaboury.sigwig.ColorUtil.*;

public class Button extends Component {
    private final Supplier<Integer> color;
    private final Supplier<String> text;
    private final Supplier<Size> size;
    private final Runnable action;
    private final Nodes icon;

    private final Signal<Boolean> mouseOver = createSignal(false);
    private final Signal<Boolean> mouseDown = createSignal(false);

    public Button(Builder builder) {
        this.color = builder.color;
        this.text = builder.text;
        this.size = builder.size;
        this.action = builder.action;
        this.icon = builder.icon;
    }

    @Override
    public Nodes render() {
           return Nodes.single(Node.builder()
                   .ref(node -> node.listen(
                           onMouseOver(e -> mouseOver.accept(true)),
                           onMouseDown(e -> mouseDown.accept(true)),
                           onMouseOut(e -> batch(() -> {
                               mouseDown.accept(false);
                               mouseOver.accept(false);
                           })),
                           onMouseUp(e -> {
                               var prev = mouseDown.get();
                               mouseDown.accept(false);
                               if (prev) {
                                   Sigui.invokeLater(action);
                               }
                           })
                   ))
                   .layout(this::layout)
                   .paint(this::paint)
                   .children(Nodes.compose(
                           icon,
                           Nodes.single(Text.line(() -> Text.basicTextLine(text.get(), fontSize()), () -> ColorUtil.contrastText(color.get())))
                   ))
                   .build());
    }

    private void layout(long yoga) {
        Yoga.YGNodeStyleSetGap(yoga, Yoga.YGGutterAll, 8f);
        Yoga.YGNodeStyleSetJustifyContent(yoga, Yoga.YGJustifyCenter);
        Yoga.YGNodeStyleSetAlignItems(yoga, Yoga.YGAlignCenter);

        switch (this.size.get()) {
            case LG -> {
                Yoga.YGNodeStyleSetHeight(yoga, 62);
                Yoga.YGNodeStyleSetPadding(yoga, Yoga.YGEdgeHorizontal, 24);
            }
            case MD -> {
                Yoga.YGNodeStyleSetHeight(yoga, 46);
                Yoga.YGNodeStyleSetPadding(yoga, Yoga.YGEdgeHorizontal, 16);
            }
            case SM -> {
                Yoga.YGNodeStyleSetHeight(yoga, 30);
                Yoga.YGNodeStyleSetPadding(yoga, Yoga.YGEdgeHorizontal, 12);
            }
            case XS -> {
                Yoga.YGNodeStyleSetHeight(yoga, 22);
                Yoga.YGNodeStyleSetPadding(yoga, Yoga.YGEdgeHorizontal, 8);
            }
        }
    }

    private float fontSize() {
        return switch (this.size.get()) {
            case LG -> 18;
            case MD, SM -> 14;
            case XS -> 12;
        };
    }

    private void paint(Canvas canvas, MetaNode node) {
        var size = node.getLayout().getSize();

        if (mouseDown.get()) {
            float pressScale = 0.95f;

            canvas.scale(pressScale, pressScale);
            canvas.translate(
                    (size.getX() * (1f - pressScale)) / 2f,
                    (size.getY() * (1f - pressScale)) / 2f
            );
        }

        try (var paint = new Paint()) {
            paint.setColor(mouseOver.get() ? hoverColor(color.get()) : color.get());
            canvas.drawRRect(Rect.makeWH(size).withRadii(8), paint);
        }
    }

    private int hoverColor(int color) {
        var oklch = oklchFromOklab(oklabFromXyz(xyzFromSrgb(srgbFromRgb(color))));
        oklch[0] = (float)Math.max(0f, Math.min(1f, oklch[0] + (oklch[0] < 0.5 ? 0.1 : -0.1)));
        return rgbFromSrgb(srgbFromXyz(xyzFromOklab(oklabFromOklch(oklch))));
    }

    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        private Supplier<Integer> color = constantSupplier(EzColors.BLUE_400);
        private Supplier<String> text = constantSupplier("");
        private Supplier<Size> size = constantSupplier(Size.MD);
        private Runnable action = () -> {};
        private Nodes icon = Nodes.empty();

        public Builder setColor(Supplier<Integer> color) {
            this.color = color;
            return this;
        }

        public Builder setColor(int color) {
            this.color = constantSupplier(color);
            return this;
        }

        public Builder setText(Supplier<String> text) {
            this.text = text;
            return this;
        }

        public Builder setText(String text) {
            this.text = constantSupplier(text);
            return this;
        }

        public Builder setSize(Supplier<Size> size) {
            this.size = size;
            return this;
        }

        public Builder setSize(Size size) {
            this.size = constantSupplier(size);
            return this;
        }

        public Builder setAction(Runnable action) {
            this.action = action;
            return this;
        }

        public Builder setIcon(Nodes icon) {
            this.icon = icon;
            return this;
        }

        public Button build() {
            return new Button(this);
        }
    }

    public enum Size {
        LG,
        MD,
        SM,
        XS
    }
}

File: ksigwig/src/main/kotlin/com/github/wilgaboury/sigwig/Button.kt
package com.github.wilgaboury.sigwig

import com.github.wilgaboury.jsignal.ReactiveUtil.batch
import com.github.wilgaboury.jsignal.ReactiveUtil.createSignal
import com.github.wilgaboury.jsignal.interfaces.SignalLike
import com.github.wilgaboury.sigui.Component
import com.github.wilgaboury.sigui.MetaNode
import com.github.wilgaboury.sigui.Nodes
import com.github.wilgaboury.sigui.SiguiUtil
import io.github.humbleui.skija.Canvas
import io.github.humbleui.skija.Paint
import io.github.humbleui.types.Rect
import com.github.wilgaboury.ksigui.listen
import com.github.wilgaboury.ksigui.node
import com.github.wilgaboury.ksigui.ref
import com.github.wilgaboury.ksigui.toNodes
import org.lwjgl.util.yoga.Yoga
import kotlin.math.max
import kotlin.math.min

class Button(
        val color: () -> Int = { EzColors.BLUE_400 },
        val text: () -> String = { "" },
        val size: () -> Size = { Size.MD },
        val action: () -> Unit = {},
        val icon: Nodes = Nodes.empty()
) : Component() {
    private val mouseOver: SignalLike<Boolean> = createSignal(false)
    private val mouseDown: SignalLike<Boolean> = createSignal(false)

    override fun render(): Nodes {
        return node {
            ref {
                tags(Button::class.qualifiedName)
                listen {
                    onMouseOver { mouseOver.accept(true) }
                    onMouseDown { mouseDown.accept(true) }
                    onMouseOut {
                        batch {
                            mouseDown.accept(false)
                            mouseOver.accept(false)
                        }
                    }
                    onMouseUp {
                        val prev = mouseDown.get()
                        mouseDown.accept(false)
                        if (prev) {
                            SiguiUtil.invokeLater(action)
                        }
                    }
                }
            }
            layout(this@Button::layout)
            paint(this@Button::paint)
            children(Nodes.compose(
                icon,
                Line(
                    { Line.basic(text(), fontSize()) },
                    { ColorUtil.contrastText(color()) }
                ).render()
            ))
        }
    }

    private fun layout(yoga: Long) {
        Yoga.YGNodeStyleSetGap(yoga, Yoga.YGGutterAll, 8f)
        Yoga.YGNodeStyleSetJustifyContent(yoga, Yoga.YGJustifyCenter)
        Yoga.YGNodeStyleSetAlignItems(yoga, Yoga.YGAlignCenter)
        when (size()) {
            Size.LG -> {
                Yoga.YGNodeStyleSetHeight(yoga, 62f)
                Yoga.YGNodeStyleSetPadding(yoga, Yoga.YGEdgeHorizontal, 24f)
            }

            Size.MD -> {
                Yoga.YGNodeStyleSetHeight(yoga, 46f)
                Yoga.YGNodeStyleSetPadding(yoga, Yoga.YGEdgeHorizontal, 16f)
            }

            Size.SM -> {
                Yoga.YGNodeStyleSetHeight(yoga, 30f)
                Yoga.YGNodeStyleSetPadding(yoga, Yoga.YGEdgeHorizontal, 12f)
            }

            Size.XS -> {
                Yoga.YGNodeStyleSetHeight(yoga, 22f)
                Yoga.YGNodeStyleSetPadding(yoga, Yoga.YGEdgeHorizontal, 8f)
            }
        }
    }

    private fun fontSize(): Float {
        return when (size()) {
            Size.LG -> 18f
            Size.MD, Size.SM -> 14f
            Size.XS -> 12f
        }
    }

    private fun paint(canvas: Canvas, node: MetaNode) {
        val size = node.layout.size
        if (mouseDown.get()) {
            val pressScale = 0.95f
            canvas.scale(pressScale, pressScale)
            canvas.translate(
                    size.x * (1f - pressScale) / 2f,
                    size.y * (1f - pressScale) / 2f
            )
        }
        Paint().use { paint ->
            paint.setColor(if (mouseOver.get()) hoverColor(color()) else color())
            canvas.drawRRect(Rect.makeWH(size).withRadii(8f), paint)
        }
    }

    private fun hoverColor(color: Int): Int {
//        var oklch = oklchFromOklab(oklabFromXyz(xyzFromSrgb(srgbFromRgb(color))));
//        oklch[0] = (float)Math.max(0f, Math.min(1f, oklch[0] + (oklch[0] < 0.5 ? 0.1 : -0.1)));
//        return rgbFromSrgb(srgbFromXyz(xyzFromOklab(oklabFromOklch(oklch))));
        val hsl = ColorUtil.hslFromRgb(color)
        hsl[2] = max(0.0, min(100.0, (hsl[2] + if (hsl[2] < 0.5) 10f else -10f).toDouble())).toFloat()
        return ColorUtil.rgbFromHsl(hsl)
    }

    enum class Size {
        LG,
        MD,
        SM,
        XS
    }
}
